src/crimson/os/futurized_objectstore.h

```diff
@@ -0,0 +1,100 @@
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <map>
+#include <typeinfo>
+#include <vector>
+
+#include "include/buffer.h"
+#include "include/buffer_fwd.h"
+#include "include/uuid.h"
+#include "osd/osd_types.h"
+
+#include <seastar/core/future.hh>
+
+namespace ceph::os{
+
+class Collection;
+class Transaction;
+
+class FuturizedObjectStore {
+
+public:
+  template <class ConcreteExceptionT>
+  class Exception : public std::logic_error {
+  public:
+    using std::logic_error::logic_error;
+
+    // Throwing an exception isn't the sole way to signalize an error
+    // with it. This approach nicely fits cold, infrequent issues but
+    // when applied to a hot one (like ENOENT on write path), it will
+    // likely hurt performance.
+    // Alternative approach for hot errors is to create exception_ptr
+    // on our own and place it in the future via make_exception_future.
+    // When ::handle_exception is called, handler would inspect stored
+    // exception whether it's hot-or-cold before rethrowing it.
+    // The main advantage is both types flow through very similar path
+    // based on future::handle_exception.
+    static bool is_class_of(const std::exception_ptr& ep) {
+      // Seastar offers hacks for making throwing lock-less but stack
+      // unwinding still can be a problem so painful to justify going
+      // with non-standard, obscure things like this one.
+      return *ep.__cxa_exception_type() == typeid(ConcreteExceptionT);
+    }
+  };
+
+  struct EnoentException : public Exception<EnoentException> {
+    using Exception<EnoentException>::Exception;
+  };
+  static FuturizedObjectStore* create(const std::string& type, const std::string& data);
+  FuturizedObjectStore() {}
+  virtual ~FuturizedObjectStore() {}
+
+  // no copying
+  explicit FuturizedObjectStore(const FuturizedObjectStore& o) = delete;
+  const FuturizedObjectStore& operator=(const FuturizedObjectStore& o) = delete;
+
+  virtual seastar::future<> mount() = 0;
+  virtual seastar::future<> umount() = 0;
+  
+  virtual seastar::future<> mkfs() = 0;
+  using CollectionRef = boost::intrusive_ptr<Collection>;
+  virtual seastar::future<bufferlist> read(CollectionRef c,
+				   const ghobject_t& oid,
+				   uint64_t offset,
+				   size_t len,
+				   uint32_t op_flags = 0) = 0;
+  virtual seastar::future<ceph::bufferptr> get_attr(CollectionRef c,
+					    const ghobject_t& oid,
+					    std::string_view name) = 0;
+
+  using attrs_t = std::map<std::string, ceph::bufferptr, std::less<>>;
+  virtual seastar::future<attrs_t> get_attrs(CollectionRef c, const ghobject_t& oid) = 0;
+  using omap_values_t = std::map<std::string,bufferlist, std::less<>>;
+  virtual seastar::future<omap_values_t> omap_get_values(
+                                         CollectionRef c,
+                                         const ghobject_t& oid,
+                                         std::vector<std::string>&& keys) = 0;
```

Comment:
May we abstract the container type for `keys`? We might want to switch for something cheaper than `std::vector` of `std::strings` in the future, I bet.

Reply:
the original ceph-osd use std::set<std::string>, seems no other type requirement.  so let's keep it as it is, if we need other type for keys, we change it later.   BTW, can we use template for virtual member function?

Reply:
> can we use template for virtual member function?

No. By *abstracting* I mean something like:
```cpp
  using omap_keys_t = std::vector<std::string>;
```
which would be quite consistent with the already existing `omap_values_t`.